### PR TITLE
Improve transpose performance with tile sm copy test=develop

### PR DESCRIPTION
**Pull request by zhaoyuchen2018**
Xlnet uses transpose heavily, the performance is poor. Use share memory to optimize memory copy.

transpose time in xlnet model:
||transpose|transpose_grad|
|--|--|--|
|before opt(ms)|227|294|
|after opt(ms)|63|165|

model performance improves ~30%

### CODE REVIEW

File: paddle/fluid/operators/transpose_op.cu [hunk 1]
```diff
@@ -0,0 +1,741 @@
...
+  constexpr bool operator()(int a, int b) const { return a == b; }
+};
+
+struct GreaterThan {
+  constexpr bool operator()(int a, int b) const { return a > b; }
+};
+
+// Value can be decided in compile time.
+template <typename FUN, int INT_32 = 32>
+constexpr bool CheckProperTileSize(int tile_long, int tile_short, int size_T,
```
Comment (wangchaochaohu) [inline paddle/fluid/operators/transpose_op.cu:46]: 这几个参数的意义是什么？

File: paddle/fluid/operators/transpose_op.cu [hunk 2]
```diff
@@ -0,0 +1,741 @@
...
+
+    if (full_tile) {
+#pragma unroll
+      for (int ind_i = x_i; ind_i < (TileX); ind_i += BlockReadRows) {
+        tile_sm[ind_i][x_j] = input[input_ind];
+        input_ind += input_inc;
+      }
+    } else {
+      if (x_j < tile_width) {
+        for (int ind_i = x_i; ind_i < (tile_height); ind_i += BlockReadRows) {
```
Comment (wangchaochaohu) [inline paddle/fluid/operators/transpose_op.cu:162]: 添加#progma unroll?

File: paddle/fluid/operators/transpose_op.cu [hunk 3]
```diff
@@ -0,0 +1,741 @@
...
+
+    if (full_tile) {
+#pragma unroll
+      for (int ind_i = x_i; ind_i < (TileY); ind_i += BlockWriteRows) {
+        output[output_ind] = tile_sm[x_j][ind_i];
+        output_ind += output_inc;
+      }
+    } else {
+      if (x_j < tile_height) {
+        for (int ind_i = x_i; ind_i < (tile_width); ind_i += BlockWriteRows) {
```
Comment (wangchaochaohu) [inline paddle/fluid/operators/transpose_op.cu:204]: 添加#pragma unroll

File: paddle/fluid/operators/transpose_op.cu (hunk 1 above)
Comment (zhaoyuchen2018) [reply paddle/fluid/operators/transpose_op.cu:46]: long和short分别表示tile的长短边, size_T表示T数据类型占多少字节

File: paddle/fluid/operators/transpose_op.cu (hunk 2 above)
Comment (zhaoyuchen2018) [reply paddle/fluid/operators/transpose_op.cu:162]: done

File: paddle/fluid/operators/transpose_op.cu (hunk 3 above)
Comment (zhaoyuchen2018) [reply paddle/fluid/operators/transpose_op.cu:204]: done

**Review by wangchaochaohu — APPROVED** [review]
LGTM

**Pull request merged**